wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileIntegrityValidator.java

```diff
@@ -0,0 +1,117 @@
+ *   <li>Project names conforms naming rules</li>
+ * </ul>
+ * </pre>
+ */
+@Singleton
+public class DevfileIntegrityValidator {
+
+  private static final Pattern PROJECT_NAME_PATTERN = Pattern.compile("^[\\w\\d]+[\\w\\d_.-]*$");
+
+  void validateDevfile(Devfile devfile) throws DevfileFormatException {
+    validateProjects(devfile);
+    Set<String> toolNames = validateTools(devfile);
+    validateCommands(devfile, toolNames);
+  }
+
+  private Set<String> validateTools(Devfile devfile) throws DevfileFormatException {
+    Set<String> existingNames = new HashSet<>();
+    Tool editorTool = null;
+    for (Tool tool : devfile.getTools()) {
+      if (!existingNames.add(tool.getName())) {
+        throw new DevfileFormatException(format("Duplicate tool name found:'%s'", tool.getName()));
+      }
+      switch (tool.getType()) {
+        case "cheEditor":
+          if (editorTool != null) {
+            throw new DevfileFormatException(
+                format(
+                    "Multiple editor tools found: '%s', '%s'",
+                    editorTool.getName(), tool.getName()));
+          }
+          editorTool = tool;
+          break;
+        case "chePlugin":
+          break;
+        default:
+          throw new DevfileFormatException(
+              format("Unsupported tool '%s' type provided:'%s'", tool.getName(), tool.getType()));
+      }
+    }
+    return existingNames;
+  }
+
+  private void validateCommands(Devfile devfile, Set<String> toolNames)
+      throws DevfileFormatException {
+    Set<String> existingNames = new HashSet<>();
+    for (Command command : devfile.getCommands()) {
+      if (!existingNames.add(command.getName())) {
+        throw new DevfileFormatException(
+            format("Duplicate command name found:'%s'", command.getName()));
+      }
+      Set<String> nonexistingToolActions =
+          command
+              .getActions()
+              .stream()
+              .map(Action::getTool)
+              .filter(t -> !toolNames.contains(t))
+              .collect(toSet());
+      if (!nonexistingToolActions.isEmpty()) {
+        throw new DevfileFormatException(
+            format(
+                "Found actions which refers to non-existing tools in command '%s':'%s'",
+                command.getName(), String.join(",", nonexistingToolActions)));
+      }
+    }
+  }
+
+  private void validateProjects(Devfile devfile) throws DevfileFormatException {
+    Set<String> existingNames = new HashSet<>();
+    for (Project project : devfile.getProjects()) {
+      if (!existingNames.add(project.getName())) {
+        throw new DevfileFormatException(
+            format("Duplicate project name found:'%s'", project.getName()));
+      }
+      if (!PROJECT_NAME_PATTERN.matcher(project.getName()).matches()) {
+        throw new DevfileFormatException(
+            format(
+                "Invalid project name found:'%s'. Name must contain only Latin letters,\n"
```

Comment:
Are you sure that error message should contain `\n`?

Reply:
Added